File: lib/utils/product_audience.dart
/// Who a product is for — the one place that decides what an audience value
/// *means*.
///
/// Pure Dart, no Flutter: the same precedent as `customer_profile_fields.dart`
/// and `size_key.dart`, so the rule is unit-testable without a widget harness.
///
/// The value is STATED by a seller, never inferred — see
/// `docs/AI/PRODUCT_AUDIENCE_PLAN.md` §1.1 for why the size band cannot stand
/// in for it (EU is unisex, so Men's and Women's share one band, and the
/// kids' band overlaps the adult one at the top).
///
/// Deliberately a SEPARATE file from `customer_profile_fields.dart`, which is
/// about the shopper's own sign-up and foot-profile fields: a product's
/// audience is a different question. The three shared string values and labels
/// are imported from there rather than re-spelled, so `'women'` and `"Women's"`
/// cannot come to mean one thing on a profile and another on a product.
library;

import 'customer_profile_fields.dart';
import 'size_key.dart';

/// The children's audience value — the one band whose sizes can be checked.
///
/// Men's and Women's share the ENTIRE adult band (EU sizing is unisex and only
/// the US/UK *chart* differs), so a size can never contradict either of them.
const String kKidsAudience = 'kids';

/// The one audience that is not a rail.
///
/// A genuinely unisex product (a plain sandal, a house slipper) is a real
/// answer, but listing it in Men's, Women's AND Kids' would put the same card
/// three times down the feed. Rail surfaces must skip it.
const String kUnisexAudience = 'unisex';

/// Every value `products.audience` may hold, in display order, with the label
/// to render for it.
///
/// The first three are `customerFootSizeCategories` — the shopper-side
/// vocabulary the AR scan and the manual picker already write
/// (`foot_measurements.shoe_category`, `profiles.foot_size_category`) — spread
/// in rather than re-declared. `unisex` is the one addition, and it exists
/// ONLY here: the shopper's own scale is never `unisex`, because a customer
/// always shops on one chart.
const List<(String, String)> productAudienceOptions = [
  ...customerFootSizeCategories,
  (kUnisexAudience, 'Unisex'),
];

/// The audiences that get a rail, in fixed order.
///
/// Explicit rather than computed as "the options minus unisex" at each call
/// site: this list is what a later phase's rails iterate and what the guard
/// test in `test/utils/product_audience_test.dart` pins — including that it
/// still matches the shopper-side vocabulary and that `unisex` is absent.
const List<String> productRailAudiences = ['men', 'women', kKidsAudience];

/// The canonical audience value for a stored string, or null.
///
/// **Exact match, deliberately.** No trimming and no case folding, because
/// this value comes from a `TEXT` column whose CHECK constraint is itself an
/// exact comparison (`audience IN ('men','women','kids','unisex')`) and from a
/// closed chip list on the seller form — so anything that is not already
/// canonical is not one of our values, and normalising it would only invent a
/// meaning the database does not hold. `'Men'` and `'unisex '` are null, not
/// `'men'` and `'unisex'`.
///
/// This also matches the existing house parser (`footSizeCategoryLabel` /
/// `savedFootSizeCategory`, which compare exactly and fall back to null), and
/// keeps the "never a guess" rule: an unrecognised, empty or malformed value
/// is null, and every caller must then render nothing rather than pick a
/// default audience.
String? productAudienceFrom(String? raw) {
  if (raw == null) return null;
  for (final (value, _) in productAudienceOptions) {
    if (value == raw) return value;
  }
  return null;
}

/// The children's band, read from the same list the size picker offers
/// (`customerKidsEuSizes`) so the warning cannot drift from what a customer is
/// actually able to enter.
final double _kidsBandMin = double.parse(customerKidsEuSizes.first);
final double _kidsBandMax = double.parse(customerKidsEuSizes.last);

/// Whether a product's sizes contradict the audience a seller picked.
///
/// Only Kids' can be contradicted: a product marked Kids' whose every stocked
/// size sits outside the children's band (EU 22–35) is almost certainly a
/// mis-tapped chip, and a soft note is worth showing. `unisex`/null claim no
/// band, so nothing can contradict them.
///
/// **Warning only, and one-directional.** This never blocks a save, never
/// rewrites the audience, and never rewrites a size — a stated value stays
/// visible and fixable rather than being reinterpreted behind the seller's
/// back (plan §0 decision 5). Callers render it as an inline note.
///
/// False unless every size is readable AND every one of them is outside the
/// band: with no sizes yet, or a custom/unparseable size in the list, there is
/// nothing solid enough to contradict, so nothing is claimed. Sizes go through
/// [sizeNumberInEu] — the same parser the rest of the app uses — so an adult
/// `'EU 42'`, a bare `'42'`, and a kids' `'EU 28'` are read the way the catalog
/// reads them.
bool audienceSizeMismatch({
  required String? audience,
  required Iterable<String> sizes,
}) {
  if (audience != kKidsAudience) return false;
  var sawASize = false;
  for (final raw in sizes) {
    final eu = sizeNumberInEu(raw);
    // One unreadable size means we do not know what this product stocks.
    if (eu == null) return false;
    // A single kids'-band size is enough to confirm the audience is right.
    if (eu >= _kidsBandMin && eu <= _kidsBandMax) return false;
    sawASize = true;
  }
  return sawASize;
}

/// The shopping scale a product's US/UK size labels are drawn on — `'men'`,
/// `'women'`, `'kids'`, or null for "no scale known".
///
/// **The one place the chart precedence lives** (plan §P3, decision #7). Call it
/// once per product and pass the result to [sizeUnitsForCategory] and to
/// `formatSize` / `displaySizeInUnit`'s `category:` — never rebuild this chain
/// at a call site, or the product page and the cart can start disagreeing about
/// which chart EU 42 is read on.
///
/// Precedence, in order:
///
///  1. **`men` / `women` on the product** → THAT product's chart, whatever scale
///     the shopper saved. A woman browsing a men's-cut shoe reads the men's
///     label, because the *item* is sold on the men's chart. This is the whole
///     change: before, the shopper's own scale won.
///  2. **`kids`** → the kids' scale, which [sizeUnitsForCategory] already
///     renders as EU-only — there is no child US/UK chart to draw on (see its
///     doc for why an offset would be a lie). Representing "EU only" as the
///     scale key is the existing convention, so nothing new is invented here.
///  3. **unset (`null`) or unrecognised** → the shopper's own scale, **exactly
///     what the product page passed before this helper existed** — including
///     the null, because a null scale is what makes `formatSize` fall back to
///     men's downstream rather than to a guessed women's chart. With every
///     product in the catalog still unset, this is the branch they all take.
///  4. **`unisex`** → the men's chart ([kDefaultSizeCategory]), the same answer
///     the app already gives when nothing is known about the shopper either.
///
/// [audienceEnabled] is the feature switch — callers pass
/// `AppConstants.productAudienceEnabled`. It is a parameter rather than a read
/// of that constant so this file stays pure Dart (no Flutter), and it is the
/// whole guard: while the switch is `false` this returns today's exact value for
/// every product, so the phase is invisible until it is turned on.
String? productSizeChart({
  required Map<String, dynamic>? product,
  required Map<String, dynamic>? profile,
  required bool audienceEnabled,
}) {
  if (!audienceEnabled) return savedFootSizeCategory(profile);
  return switch (productAudienceFrom(product?['audience']?.toString())) {
    'men' => kDefaultSizeCategory,
    'women' => 'women',
    kKidsAudience => kKidsAudience,
    kUnisexAudience => kDefaultSizeCategory,
    // Null and anything unrecognised: the shopper's own scale, verbatim.
    _ => savedFootSizeCategory(profile),
  };
}

/// The label to render for a canonical audience value, or null.
///
/// Null for null, for an unrecognised value, and for anything that came back
/// from `productAudienceFrom` as null — a caller that has a label is
/// guaranteed to have a real audience.
String? productAudienceLabel(String? value) {
  if (value == null) return null;
  for (final (key, label) in productAudienceOptions) {
    if (key == value) return label;
  }
  return null;
}

File: customer-portal/src/lib/productAudience.js
/**
 * Product audience — who a product is for, and the one place that decides what
 * an audience value *means*. A port of `lib/utils/product_audience.dart`.
 *
 * ## Why the value is stated rather than derived
 *
 * The obvious shortcut is to read the audience off the sizes a product stocks.
 * It cannot be done: EU sizing is unisex, so Men's and Women's share one entire
 * band, and the kids' band overlaps the adult one at the top (EU 35 is both a
 * child's size and the bottom of the women's scale). Two of the four values
 * would be guessed wrong. So `products.audience` is a value a seller picks, and
 * an unset one stays unset.
 *
 * ## The vocabulary is the shopper's, plus one
 *
 * Men's / Women's / Kids' are `FOOT_SIZE_CATEGORIES` — the same three scales the
 * size picker and the AR scan write to a profile. Spreading them in rather than
 * re-spelling them is what stops `'women'` from meaning one thing on a profile
 * and another on a product. `unisex` is the one addition, and it exists only
 * here: a shopper always shops on one chart, so their own scale is never unisex.
 *
 * ## Null is an answer, never a default
 *
 * `null` means "the seller has not said", and nothing reads it as "any
 * audience". A product with no audience stays in the catalog and in search; it
 * simply appears in no audience rail. `parse` is an exact match for the same
 * reason the column's CHECK constraint is — inventing a meaning for `'Men'` would
 * be storing a value the database does not hold.
 */
import { EU_KIDS_SIZES, FOOT_SIZE_CATEGORIES } from './footRules.js'
import { sizeNumberInEu } from './sizeKeyRules.js'

/** The children's audience value — the one band whose sizes can be checked. */
export const KIDS_AUDIENCE = 'kids'

/**
 * The one audience that is not a rail.
 *
 * A genuinely unisex product is a real answer, but listing it in Men's, Women's
 * AND Kids' would put the same card three times down the feed, so rail surfaces
 * must skip it.
 */
export const UNISEX_AUDIENCE = 'unisex'

/** Every value `products.audience` may hold, in display order. */
export const PRODUCT_AUDIENCE_OPTIONS = [
  ...FOOT_SIZE_CATEGORIES,
  { value: UNISEX_AUDIENCE, label: 'Unisex' },
]

/** The audiences that get a rail, in fixed order — `unisex` is absent by design. */
export const PRODUCT_RAIL_AUDIENCES = ['men', 'women', KIDS_AUDIENCE]

/**
 * The canonical audience value for a stored string, or null.
 *
 * **Exact match, deliberately.** No trimming and no case folding: the column's
 * CHECK constraint is itself an exact comparison, so anything that is not
 * already canonical is not one of our values. `'Men'` and `'unisex '` are null,
 * not `'men'` and `'unisex'`.
 */
export function productAudienceFrom(raw) {
  for (const option of PRODUCT_AUDIENCE_OPTIONS) {
    if (option.value === raw) return option.value
  }
  return null
}

/**
 * The label to render for a canonical value, or null.
 *
 * Null for null and for anything unrecognised, so a caller that has a label is
 * guaranteed to have a real audience — a chip group never has to decide what to
 * draw for a value it does not know.
 */
export function productAudienceLabel(value) {
  if (value === null || value === undefined) return null
  for (const option of PRODUCT_AUDIENCE_OPTIONS) {
    if (option.value === value) return option.label
  }
  return null
}

/** The kids' band, read from the list the size picker offers. */
const KIDS_BAND_MIN = Number.parseFloat(EU_KIDS_SIZES[0])
const KIDS_BAND_MAX = Number.parseFloat(EU_KIDS_SIZES[EU_KIDS_SIZES.length - 1])

/**
 * Whether a product's sizes contradict the audience a seller picked.
 *
 * Only Kids' can be contradicted: a product marked Kids' whose every stocked
 * size sits outside the children's band is almost certainly a mis-tapped chip,
 * and a soft note is worth showing. `unisex` and null claim no band, so nothing
 * can contradict them.
 *
 * **Warning only, and one-directional.** This never blocks a save, never
 * rewrites the audience and never rewrites a size — a stated value stays visible
 * and fixable rather than being reinterpreted behind the seller's back.
 *
 * False unless *every* size is readable AND every one of them is outside the
 * band: with no sizes yet, or one custom/unreadable size in the list, there is
 * nothing solid enough to contradict, so nothing is claimed. A single size
 * inside the band is enough to confirm the audience is right.
 */
export function audienceSizeMismatch({ audience, sizes }) {
  if (audience !== KIDS_AUDIENCE) return false

  let sawASize = false
  for (const raw of sizes ?? []) {
    const eu = sizeNumberInEu(raw)
    if (eu === null) return false
    if (eu >= KIDS_BAND_MIN && eu <= KIDS_BAND_MAX) return false
    sawASize = true
  }
  return sawASize
}
